rabbitmq/connection.go

```diff
@@ -0,0 +1,126 @@
+package rabbitmq
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+// RabbitMQConnection handles RabbitMQ connection management with automatic reconnection.
+type RabbitMQConnection struct {
+	url       string
+	conn      *amqp.Connection
+	connClose chan *amqp.Error
+	mu        sync.RWMutex // Protects conn and connClose fields
+}
+
+func NewRabbitMQConnection(url string) *RabbitMQConnection {
+	return &RabbitMQConnection{
+		url: url,
+	}
+}
+
+func (cm *RabbitMQConnection) Connect(ctx context.Context, maxRetries int) (*amqp.Connection, error) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	if cm.conn != nil && cm.isConnectionAlive() {
+		return cm.conn, nil // Already connected and alive
+	}
+
+	// Close existing connection if any
+	if cm.conn != nil {
+		cm.conn.Close()
+		cm.conn = nil
+		cm.connClose = nil
+	}
+
+	// Create new connection
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// Apply backoff delay (except for first attempt)
+		if attempt > 0 {
+			backoff := nextBackoff(attempt - 1)
+			time.Sleep(backoff)
+		}
+
+		// Try to create connection
+		conn, err := amqp.Dial(cm.url)
+		if err != nil {
+			if attempt == maxRetries-1 {
+				return nil, fmt.Errorf("failed to connect after %d attempts: %w", maxRetries, err)
+			}
+			continue
+		}
+
+		// Connection created successfully
+		cm.conn = conn
+		cm.connClose = make(chan *amqp.Error, 1)
+		conn.NotifyClose(cm.connClose)
+		return cm.conn, nil
+	}
+
+	return nil, fmt.Errorf("exceeded maximum connection attempts")
```

Comment:
In which case do we reach this return?
Looks like we'll always return from line 60 in case we can't establish a connection.

Maybe we could declare `err` outside the loop, add it to this message and remove line 60.